### ISSUE TITLE
spec: problem parsing named channels

### ISSUE THREAD

**Issue by robpike** [golang/go#21953]
Consider this program:

```
package main

import (
	"fmt"
)

type Foo chan string

func main() {
	c := make(Foo, 1)
	c <- "gotcha"
	fmt.Println(<-c)
}

func bar(my <-Foo) {
}
```

All is well. Now try to declare a direction for a Foo:

```
package main

import (
	"fmt"
)

type Foo chan string

func main() {
	var x <-Foo
}

```

**Comment by cespare** [comment]
Isn't this kind of like doing

```
type MyInt int
var x uMyInt // Is this a uint?
```

i.e., I wouldn't expect this to work at all.

**Comment by dsnet** [comment]
I'm tentatively marking this as a language change, unless you are suggesting that the current parser is not matching the written spec (in which case this is a bug).

**Comment by robpike** [comment]
I think it's a language change.

**Comment by griesemer** [comment]
Definitively a language change, possibly backward compatible.

But why would we want to allow such a thing? It makes the direction arrow an operator on types, and only on channel types. And it only can make a bidirectional channel a read-only channel. What if I wanted to create a write-only channel? In the current design of the language `<-` arrow is really part of the channel syntax pattern, not a separate operator when it comes to channel types. The syntax is simply not compositional (we cannot interject a `<-` into the middle of a channel type).

I don't see why this would be worth the effort.

**Comment by leonklingele** [comment]
This is requesting similar syntax to https://github.com/golang/go/issues/22189.

**Comment by robpike** [comment]
I'm not saying it is or is not worth the effort, but it definitely a flaw in the design that one can declare read-only channel types only if they are unnamed. I can't think of another property of the language that works like this, for instance that changes the possible types acceptable by a function.

**Comment by ianlancetaylor** [comment]
See also #23415.

**Comment by mccolljr** [comment]
Take a look at this playground example: https://play.golang.org/p/ih1Uxm4Gfna
```go
package main

import (
	"fmt"
)

type (
	Foo         chan string
	FooReadable <-chan string
	FooWritable chan<- string
)

func Send(f FooWritable, msg string) {
	f <- msg
}

func Recv(f FooReadable) string {
	return <-f
}

func Send2(f chan<- string, msg string) {
	f <- msg
}

func Recv2(f <-chan string) string {
	return <-f
}

func main() {
	Scenario1()
	Scenario2()
	Scenario3()
	Scenario4()
	Scenario5()
}

func Scenario1() {
	// works
	//
	foo := make(chan string)

	// compile error
	//
	// foo := make(Foo)

	go Send(foo, "message")
	fmt.Println(Recv(foo))
}

func Scenario2() {
	// works
	//
	// foo := make(chan string)

	// works as well
	//
	foo := make(Foo)

	go Send2(foo, "message2")
	fmt.Println(Recv2(foo))
}

func Scenario3() {
	foo := make(chan string)

	reader := FooReadable(foo)
	writer := FooWritable(foo)

	go Send(writer, "message3")
	fmt.Println(Recv(reader))

	go Send2(writer, "message4")
	fmt.Println(Recv2(reader))
}

func Scenario4() {
	foo := make(Foo)

	reader := (<-chan string)(foo)
	writer := (chan<- string)(foo)

	go Send(writer, "message5")
	fmt.Println(Recv(reader))

	go Send2(writer, "message6")
	fmt.Println(Recv2(reader))
}

func Scenario5() {
	temp := make(chan string)
	foo := Foo(temp)

	reader := (<-chan string)(foo)
	writer := (chan<- string)(foo)

	go Send(writer, "message7")
	fmt.Println(Recv(reader))

	go Send2(writer, "message8")
	fmt.Println(Recv2(reader))
}
```

Is this expected behavior? It seems odd to me that type conversion can be done in one direction but not the other:

1. `Foo` can become `chan string`, `<-chan string` and `chan<- string`
2. `chan string` can become `Foo`, `FooReadable` and `FooWritable`
3. `<-chan string` can become `FooReadable`
4. `chan<- string` can become `FooWritable`
5. `Foo` **cannot become** `FooReadable` or `FooWritable`

Is the inability of `Foo` to be converted in this way intentional?

**Comment by ianlancetaylor** [comment]
@mccolljr I don't think this proposal is the right place to discuss the details of Go type system.  For the reasons for what you are seeing, see https://golang.org/ref/spec#Assignability.  If you want to discuss this, see https://golang.org/wiki/Questions.  If you want to propose a change, please use a different issue.  Thanks.

**Comment by ianlancetaylor** [comment]
@mccolljr I apologize for making you feel that I was talking down to you.  That was in no way my intent.

That said, please keep this issue for discussion of the specific proposal above.  Discussions of type assignability rules are simply not related to this proposal, which is not about assignability, or type conversions, at all.  Thanks for your consideration.

**Comment by griesemer** [comment]
An possible problem with this proposal might be ambiguity of the parse tree (requiring type information to resolve it). To be investigated.

**Comment by pmaddams** [comment]
Hi folks. I think this proposal is a great idea; I came up with the following use case: https://play.golang.org/p/ktT0Mhgw8z8

```
// Public domain.

package main

import (
	"fmt"
	"math/big"
)

var (
	one = big.NewInt(1)
	ten = big.NewInt(10)
)

// stream represents a stream of arbitrary-precision integers.
type stream chan *big.Int

// values returns a stream of values in [lo, hi]. If hi is nil, the stream is infinite.
func values(lo, hi *big.Int) stream {
	st := make(stream)
	go func() {
		z := new(big.Int).Set(lo)
		for {
			if hi != nil && z.Cmp(hi) > 0 {
				break
			}
			st <- new(big.Int).Set(z)
			z.Add(z, one)
		}
		close(st)
	}()
	return st
}

func main() {
	for z := range values(one, ten) {
		fmt.Println(z)
	}
}
```

It would be nice to impose the additional constraint that the channel returned by “values” is receive-only. As is, because I want type safety, I’ve given up on the “stream” abstraction and have a family of functions that return <-chan *big.Int.
